feat: explain the limitations of serialization (#10944)
withastro/docs: src/content/docs/en/guides/framework-components.mdx

</div>
```

> :::caution
> You can pass a function as a prop to a framework component, but this only works during server rendering. If you try to use the function in a hydrated component (for example, as an event handler), an error will occur.
> 
> This is because functions cannot be _serialized_ (transferred from the server to the client) by Astro. Objects with circular references are also not serializable.
> 
> Props provided to hydrated components must be serializable. The following prop types are supported: 
> `plain object`, `number`, `string`, `Array`, `Map`, `Set`, `RegExp`, `Date`, `BigInt`, `URL`, `Uint8Array`, `Uint16Array`, `Uint32Array`, and `Infinity`
> :::

Reviewer: ```suggestion

Props that are passed to framework functions must be [serialized](https://developer.mozilla.org/en-US/docs/Glossary/Serialization): translated into a format suitable for transfer over a network, or storage. However, for performance and/or security reasons, Astro does not serialize every type of data structure. Therefore, there are some limitations on what can be passed as props.

Notably, you can only pass a function as a prop to a framework component during server rendering. If you try to use the function in a hydrated component (for example, as an event handler), an error will occur. Similarly, passing an object with circular reference is not supported.

The following prop types are supported: 
`plain object`, `number`, `string`, `Array`, `Map`, `Set`, `RegExp`, `Date`, `BigInt`, `URL`, `Uint8Array`, `Uint16Array`, `Uint32Array`, and `Infinity`
```
OK, after some chat with the maintainers, this is my first pass at more generic docs! Will ask people to weigh in and poke holes in this, tell me exactly how it's wrong so we can make it better! 🙌 

cc @Fryuni @martrapp @OliverSpeir 

(PS I'm not necessarily thrilled with the existing content (middle paragraph here) that we're incorporating, so improving the old content that was here is fair game, too!)

Author: Great! Thank you so much for doing this. I'll take some time to look into rewriting the existing content as well!

</div>
```

> Props that are passed to framework functions must be [serialized](https://developer.mozilla.org/en-US/docs/Glossary/Serialization): translated into a format suitable for transfer over a network, or storage. However, for performance and/or security reasons, Astro does not serialize every type of data structure. Therefore, there are some limitations on what can be passed as props.

Author: Thank you, Sarah! This looks a lot better, I appreciate the help! I Good call out on the contradiction between paragraph 1 and 3.

I think there's something I missed, though. 

For paragraph 1, I believe we want to say the props *should* be *serializable*. The distinction being that there's no action a user needs to take (which saying it must be serialized implies). The props either are able to be serialized, or they are not. The limitation is inherent with the data type. For example, a function is not serializable, therefore it cannot be passed to interactive client components like you mentioned.

Let me update the PR with some changes, and incorporate the ones you've added.

> Props that are passed to framework functions must be [serialized](https://developer.mozilla.org/en-US/docs/Glossary/Serialization): translated into a format suitable for transfer over a network, or storage. However, for performance and/or security reasons, Astro does not serialize every type of data structure. Therefore, there are some limitations on what can be passed as props.

Reviewer: > For paragraph 1, I believe we want to say the props should be serializable. 

This was the distinction that our maintainers discussed to move away from that wording! Apparently, functions *can* be serialized, it's just that Astro chooses not to. That was the basis of my wording re: Astro doesn't serialize everything. So I don't think that's a change that should be made

> Props that are passed to framework functions must be [serialized](https://developer.mozilla.org/en-US/docs/Glossary/Serialization): translated into a format suitable for transfer over a network, or storage. However, for performance and/or security reasons, Astro does not serialize every type of data structure. Therefore, there are some limitations on what can be passed as props.

Author: Hi Sarah! My apologies the misunderstanding, that makes sense! I really appreciate the feedback. With that said, I believe what you've added looks great!

The following prop types are supported: 
`plain object`, `number`, `string`, `Array`, `Map`, `Set`, `RegExp`, `Date`, `BigInt`, `URL`, `Uint8Array`, `Uint16Array`, `Uint32Array`, and `Infinity`

> Props that are not serializable can still be passed to a framework components during server rendering. Framework components that are hydrated require that props are serializable.

Author: I think this is largely accurate. I would say maybe instead of `Non-supported data structures` that we could say `non-serializable data structures`, just so that there's no ambiguity.

> Props that are not serializable can still be passed to a framework components during server rendering. Framework components that are hydrated require that props are serializable.

Reviewer: Again, see my point above: even functions ARE serializable (if you're determined enough 😄 ), it's that Astro chooses not to support serializing them. So again, that's not the nuance I was given by our maintainers and the basis for my wording it that way,

</div>
```

> Props that are passed to interactive framework components [using a `client:*` directive](/en/reference/directives-reference/#client-directives) must be [serialized](https://developer.mozilla.org/en-US/docs/Glossary/Serialization): translated into a format suitable for transfer over a network, or storage. However Astro does not serialize every type of data structure. Therefore, there are some limitations on what can be passed as props to hydrated components.

Reviewer 2: comma nit
```suggestion
Props that are passed to interactive framework components [using a `client:*` directive](/en/reference/directives-reference/#client-directives) must be [serialized](https://developer.mozilla.org/en-US/docs/Glossary/Serialization): translated into a format suitable for transfer over a network, or storage. However, Astro does not serialize every type of data structure. Therefore, there are some limitations on what can be passed as props to hydrated components.
```

The following prop types are supported: 

> `plain object`, `number`, `string`, `Array`, `Map`, `Set`, `RegExp`, `Date`, `BigInt`, `URL`, `Uint8Array`, `Uint16Array`, `Uint32Array`, and `Infinity`

Reviewer 3: ```suggestion
plain object, `number`, `string`, `Array`, `Map`, `Set`, `RegExp`, `Date`, `BigInt`, `URL`, `Uint8Array`, `Uint16Array`, `Uint32Array`, and `Infinity`
```
The others are actual types, but "plain object" isn't. It could be "plain `object`" but I think it's better without it.

> Non-supported data structures passed to components, such as functions, can only be used during the component's server rendering and cannot be used to provide interactivity. For example, passing functions to hydrated components is not supported because Astro cannot pass functions from the server in a way that make them executable on the client.

Reviewer 2: ````suggestion
Non-supported data structures passed to components, such as functions, can only be used during the component's server rendering and cannot be used to provide interactivity. For example, passing functions to hydrated components is not supported because Astro cannot pass functions from the server in a way that makes them executable on the client.

```
